### PR TITLE
implicit casts and add missing docs

### PR DESCRIPTION
**Pull request by dnfield**
implicit casts is for NNBD

I also added missing docs to a few places.  I'm skipping the ones on window.dart because I'm going to open a separate PR to make those private.

### CODE REVIEW

File: lib/ui/lerp.dart [hunk 1]
```diff
@@ -10,5 +10,5 @@ double lerpDouble(num a, num b, double t) {
     return null;
   a ??= 0.0;
   b ??= 0.0;
-  return a + (b - a) * t;
+  return a + (b - a) * t as double;
```
Comment (gspencergoog) [inline lib/ui/lerp.dart]: Maybe some extra parens so I don't have to think about precedence?
```suggestion
  return (a + (b - a) * t) as double;
```

File: lib/web_ui/lib/src/ui/canvas.dart [hunk 2]
```diff
@@ -953,11 +980,18 @@ class Canvas {
...
-  //
-  //  * [drawRawAtlas], which takes its arguments as typed data lists rather
-  //    than objects.
+  /// Draws part of an image - the [atlas] - onto the canvas.
+  ///
+  /// This method allows for optimization when you only want to draw part of an
+  /// image on the canvas, such as when using sprites or zooming. It is more
+  /// efficient than using clips or masks directly.
+  ///
+  /// All parameters mmust not be null.
```
Comment (gspencergoog) [inline lib/web_ui/lib/src/ui/canvas.dart]:
```suggestion
  /// All parameters must not be null.
```

File: lib/ui/geometry.dart [hunk 3]
```diff
@@ -81,11 +81,9 @@ abstract class OffsetBase {
...
   @override
   bool operator ==(dynamic other) {
-    if (other is! OffsetBase)
-      return false;
-    final OffsetBase typedOther = other;
-    return _dx == typedOther._dx &&
-           _dy == typedOther._dy;
+    return other is OffsetBase
+           && other._dx == _dx
+           && other._dy == _dy;
```
Comment (a14n) [inline lib/ui/geometry.dart]: nit: indentation

```dart
    return other is OffsetBase
        && other._dx == _dx
        && other._dy == _dy;
```

File: lib/ui/painting.dart [hunk 4]
```diff
@@ -2791,19 +2781,10 @@ class ImageFilter {
...
-      return false;
-    }
-    if (!_listEquals<double>(_data, typedOther._data)) {
-      return false;
-    }
-
-    return _filterQuality == typedOther._filterQuality;
+    return other is ImageFilter
+        && other._type == _type
+        && _listEquals(other._data, _data)
```
Comment (a14n) [inline lib/ui/painting.dart]: `_listEquals<double>` ?

File: lib/ui/lerp.dart (hunk 1 above)
Comment (dnfield) [reply lib/ui/lerp.dart]: I did it this way at first but it triggered the unnecessary parens lint.

File: lib/ui/geometry.dart (hunk 3 above)
Comment (dnfield) [reply lib/ui/geometry.dart]: Done

File: lib/ui/painting.dart (hunk 4 above)
Comment (dnfield) [reply lib/ui/painting.dart]: Done

File: lib/ui/lerp.dart (hunk 1 above)
Comment (gspencergoog) [reply lib/ui/lerp.dart]: Well, that's too bad.  Almost seems like a bug in that lint: sure, it's clearly not strictly necessary, but neither is it clear to me that the `as double` applies to the whole thing as opposed to the `t` alone (which is already a double).

Anyhow, it's not worth putting in an "ignore:" so I'm fine with just leaving the parens off in this case.

**Comment by MatthewPMatherne7** [comment]
chime bank

**Comment by MatthewPMatherne7** [comment]
[mmaction2-0.5.0.tar.zip](https://github.com/flutter/engine/files/5122019/mmaction2-0.5.0.tar.zip)

**Pull request merged**